docs/rfcs/0004-agent-flow.md

```diff
@@ -0,0 +1,166 @@
+
+Goals of this design document
+
+* Define high level concepts and goals
+* Define high level concepts of the execution path
+
+This document represents ideals and not technical implementation. 
+
+Non Goals of this design document
+
+* Define a technical implementation or configuration implementation
+
+# 2. Broad Solution Path
+
+Conversation around should the components be assembled via message passing, via expressions, or a hybrid approach.
+
+**Note: Consider all examples pseudoconfig**
+
+## 2.1 Expression Based
+
+Expression based is writing expressions that allow referencing other components streams/outputs/values and using them directly. Expressions allow referencing other fields, along with complex programming concepts. (functions, arithmetic ect). For instance `field1 = len(service_discover1.targets)`.
+
+**Pros**
+
+* Easier to Implement, evaluating expressions can map directly to existing config structs
+* Components are more reusable, you can pass basic types around (string, int, bool) in addition to custom types
+
+**Cons**
+* Harder for users to wire things together
+  * References to components are more complex, which may be harder to understand 
+* Harder to build a GUI for
+  * Every field of a component is potentially dynamic, making it harder to represent visually
+
+
+## 2.2 Message Based
+
+Message based is where components have no knowledge of other components and information is passed strictly via input and output streams. 
+
+**Pros**
+
+* Easier for users to understand the dependencies between components
+* Easier to build a GUI for
+    * Inputs and Outputs are well typed
+    * Messages mean users are only dragging outputs to inputs. The components themselves determine how to interpret the input needing less configuration from the user.
+* References between components are no more than strings, making the text-based representation language agnostic (e.g., it could be YAML, JSON, or any language)  
+
+**Cons**
+
+* More time consuming to implement, existing integrations/items would need to be componentized
+* Larger type system needed
+* More structured to keep the amount of types down
+
+Messages would require a more rigid and well defined type structure. For instance for getting credentials from various sources and passing those credentials around we would want to avoid the following.
+
+* MySQLCredentials
+* RedisCredentials
+* RemoteWriteCredentials
+* MySQLCredentialsSourceS3
+* MySQLCredentialsSourceVault
+* MySQLCredentialsSourceConsul
+* RedisCredentialsSourceS3
+* RedisCredentialsSourceVault
+* RedisCredentialsSourceConsul
+* RemoteWriteCredentialsSourceS3
+* RemoteWriteCredentialsSourceVault
+* RemoteWriteCredentialsSourceConsul
+
+And instead have one component that can read from many sources and outputs a single `Credential Type` and its up the destination component to intepret that correctly. 
```

Comment:
```suggestion
Messages require a more rigid type structure to minimize the number of total components. 

For example, it would be preferable to have a single `Credential` type that can be emitted by an s3, Vault, or Consul component. These components would then need to set a field that marks their output as a specific kind of Credential (such as Basic Auth or Bearer Auth).

If, instead, you had multiple Credential types, like `MySQLCredentials` and `RedisCredentials`, you would have the following components:

* Vault component for MySQL credentials 
* Vault component for Redis credentials 
* S3 component for MySQL credentials 
* S3 component for Redis credentials 
* (and so on)
```